packages/flutter/test/material/tabs_test.dart

```diff
@@ -2580,4 +2580,81 @@ void main() {
     await tester.pumpAndSettle();
     pageController.removeListener(pageControllerListener);
   });
+
+  testWidgets('onTabChanged responds to Tab click and Scroll Notifications', (WidgetTester tester) async {
+    const List<Tab> tabs = <Tab>[
+      Tab(text: 'Zeroth'), Tab(text: 'First'), Tab(text: 'Second')
+    ];
+    final List<int> logs = <int>[];
+
+    await tester.pumpWidget(MaterialApp(
+        home: DefaultTabController(
+          length: tabs.length,
+          child: Scaffold(
+            appBar: AppBar(
+              bottom: const TabBar(tabs: tabs),
+            ),
+            body: TabBarView(
+              children: tabs.map((Tab tab) => Center(child: Text('${tab.text} Tab'))).toList(),
+              onTabChanged: (int tabNumber) {
+                logs.add(tabNumber);
+              },
+            ),
+          ),
+        )));
+
+    // Initially logs should be empty.
+    expect(logs.isEmpty, isTrue);
+
+    // Change the Tab through TabBar.
+    await tester.tap(find.text('Second'));
+    await tester.pumpAndSettle();
+
+    expect(logs.first, 2);
+    logs.removeLast();
+
+    // Change the Tab through swipe gesture.
+    await tester.fling(find.text('Second Tab'), const Offset(400, 0), 800);
+    await tester.pumpAndSettle();
+
+    expect(logs.first, 1);
+  });
+
+  testWidgets('onTabChanged is called only when tab is changed', (WidgetTester tester) async {
+    const List<Tab> tabs = <Tab>[
+      Tab(text: 'Zeroth'), Tab(text: 'First'), Tab(text: 'Second')
+    ];
+    final List<int> logs = <int>[];
+
+    // This test indicates that, even if user is scrolling, the onTabChanged is
+    // called only when current page is changed.
+    await tester.pumpWidget(MaterialApp(
+        home: DefaultTabController(
+          length: tabs.length,
+          child: Scaffold(
+            appBar: AppBar(
+              bottom: const TabBar(tabs: tabs),
+            ),
+            body: TabBarView(
+              children: tabs.map((Tab tab) => Center(child: Text('${tab.text} Tab'))).toList(),
+              onTabChanged: (int tabNumber) {
+                logs.add(tabNumber);
+              },
+            ),
+          ),
+        )));
+
+    // The default viewport being Size(800, 600), scrolling the screen by
+    // Offset(-380, 0) won't change the tab. The tab will be changed once
+    // the Offset.dx is greater than -400.
+    await tester.drag(find.text('Zeroth Tab'), const Offset(-380, 0));
+    await tester.pumpAndSettle();
+
+    expect(logs.isEmpty, isTrue);
+
+    await tester.drag(find.text('Zeroth Tab'), const Offset(-410, 0));
+    await tester.pumpAndSettle();
+
+    expect(logs.first, 1);
```

Comment:
This section probably isn't necessary, since we test for it in the previous test.